Work in Progress: migration to Cassandra java driver v4

Nowheresly opened the pull request:
Hi @trask 

this is a PR to migrate to the last version of the java driver for cassandra.
Unfortunately, the last version (v4) is *NOT* binary compatible with the previous version. See here:
https://github.com/datastax/java-driver/tree/4.x/upgrade_guide

This new version requires java 8+ and radically changes several aspects such as:
- the way asynchronous is managed
- the way the driver is configured

This PR (still work in progress) is my first attempt to get a working version of glowroot-central with as less refactoring as possible to ease the migration (and the review!). To make this possible, I had to replace some asynchronous calls with synchronous ones.

I work on some additional commits to refactor the asynchronous part (and revert back to asynchronous what I temporary changed) to leverage the new features of the v4 driver. We can expect a lot of performance and scalability improvements.

I'll come with more explanations about what's inside this PR.

trask commented on central/src/main/java/org/glowroot/central/repo/ActiveAgentDao.java:
```diff
@@ -121,9 +118,9 @@
-        boundStatement.setTimestamp(0, new Date(from));
-        boundStatement.setTimestamp(1, new Date(revisedTo));
+        BoundStatement boundStatement = readTopLevelPS.get(rollupLevel).bind()
+            .setInstant(0, new Date(from).toInstant())
```
```suggestion
            .setInstant(0, Instant.ofEpochMilli(from))
```

Nowheresly replied:
Sure ! that's sound more than legit. 
Done

trask reviewed (commented):
I haven't gotten through all of it, but looks really great so far

Nowheresly commented on central/src/main/java/org/glowroot/central/repo/CentralConfigDao.java:
```diff
@@ -94,14 +91,14 @@ void write(String key, Object config, String priorVersion) throws Exception {
+            .setString(i++, newValue)
+            .setString(i++, key)
+            .setString(i++, currValue);
+        AsyncResultSet asyncresults = session.update(boundStatement);
```
We can't rename this variable to results, since result is already used in the same scope with a different type (ResultSet instead of AsyncResultSet).
This is here a "false" asynchronous call, since the ".get()" method is called in `session.update(boundStatement)`.

I'm pretty sure all these kind of blocking calls could be replaced with some efficient non-blocking calls with `CompletionStage`.

This will come later :)

Nowheresly commented on central/src/main/java/org/glowroot/central/repo/IncidentDao.java:
```diff
@@ -183,45 +183,45 @@ public void resolveIncident(OpenIncident openIncident, long resolveTime) throws
-        boundStatement.setBytes(i++, notificationBytes);
-        boundStatement.setTimestamp(i++, new Date(openIncident.openTime()));
-        boundStatement.setInt(i++, adjustedTTL);
+        boundStatement = boundStatement.setByteBuffer(i++, notificationBytes)
```
ok

Nowheresly commented on webdriver-tests/src/test/java/org/glowroot/tests/CassandraWrapper.java:
```diff
@@ -20,53 +20,67 @@
             CASSANDRA_JVM = System.getProperty("cassandra.jvm");
             if (Strings.isNullOrEmpty(CASSANDRA_JVM)) {
-                throw new IllegalStateException("Cassandra 2.x itself requires Java 7 or Java 8,"
+                throw new IllegalStateException("Java version betwwen 8 and 11 is required"
```
yes, sorry

Nowheresly commented on central/src/main/java/org/glowroot/central/repo/GaugeValueDaoImpl.java:
```diff
@@ -503,17 +500,17 @@ private ListenableFuture<?> rollupOneFromRows(int rollupLevel, String agentRollu
-        boundStatement.setDouble(i++, totalWeightedValue / totalWeight);
-        boundStatement.setLong(i++, totalWeight);
-        boundStatement.setInt(i++, adjustedTTL);
+        boundStatement = boundStatement.setDouble(i++, totalWeightedValue / totalWeight)
```
Yes, sure it is nice to read. But conditioning parameters break this chaining. I don't know yet what's the idiomatic way to do that...

Nowheresly commented:
Hi @trask
first of all, thank you for taking the time to review this PR :)
This is very much appreciated.
This PR is quite big because of the numerous breaking changes introduced in the new cassandra driver.

As explained previously, in order to get a working PR without too much refactoring, I had to replace asynchronous calls with synchronous ones.

I have now added another [commit](https://github.com/glowroot/glowroot/pull/923/commits/1fbb8848af1fc015faa3b8513adb46ec057bdc4a) to this PR that reverts back the asynchronous calls.

The new driver introduced a new interface `AsyncResultSet` in addition to the well-known `ResultSet`.

This design choice is clearly explained in the [documentation](https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/paging/#asynchronous-paging).

> In previous versions of the driver, the synchronous and asynchronous APIs returned the same ResultSet type. This made asynchronous paging very tricky, because it was very easy to accidentally trigger background synchronous queries (which would defeat the whole purpose of async, and potentially introduce deadlocks).

So, in this new commit, I had to replace `ListenableFuture` with `CompletionStage` and/or `CompletionFuture`, then to replace `ResultSet` with `AsyncResultSet` where needed and adjust the way paging is performed to comply with the `AsyncResultSet` interface.

Here is an example:

Before:

```
    private ListenableFuture<?> rollupOverallSummaryFromRows(RollupParams rollup,
            AggregateQuery query, Iterable<Row> rows) throws Exception {
        double totalDurationNanos = 0;
        double totalCpuNanos = 0;
        double totalAllocatedBytes = 0;
        long transactionCount = 0;
        for (Row row : rows) {
            totalDurationNanos += row.getDouble(0);
            totalCpuNanos += row.getDouble(1);
            totalAllocatedBytes += row.getDouble(2);
            transactionCount += row.getLong(3);
        }
        int i = 0;
        BoundStatement boundStatement =
                getInsertOverallPS(summaryTable, rollup.rollupLevel()).bind()
            .setString(i++, rollup.agentRollupId())
            .setString(i++, query.transactionType())
            .setInstant(i++, Instant.ofEpochMilli(query.to()))
            .setDouble(i++, totalDurationNanos)
            .setDouble(i++, totalCpuNanos)
            .setDouble(i++, totalAllocatedBytes)
            .setLong(i++, transactionCount)
            .setInt(i++, rollup.adjustedTTL().generalTTL());
        return session.writeAsync(boundStatement);
    }
```

After the migration:

```
    private CompletableFuture<?> rollupOverallSummaryFromRows(RollupParams rollup,
            AggregateQuery query, List<AsyncResultSet> results) {
        DoubleAccumulator totalDurationNanos = new DoubleAccumulator(Double::sum, 0.0);
        DoubleAccumulator totalCpuNanos = new DoubleAccumulator(Double::sum, 0.0);
        DoubleAccumulator totalAllocatedBytes = new DoubleAccumulator(Double::sum, 0.0);
        AtomicLong transactionCount = new AtomicLong(0);

        Function<AsyncResultSet, CompletableFuture<?>> compute =
                new Function<AsyncResultSet, CompletableFuture<?>>(){
            @Override
            public CompletableFuture<?> apply(AsyncResultSet result) {
            for (Row row : result.currentPage()) {
                totalDurationNanos.accumulate(row.getDouble(0));
                totalCpuNanos.accumulate(row.getDouble(1));
                totalAllocatedBytes.accumulate(row.getDouble(2));
                transactionCount.addAndGet(row.getLong(3));
            }
            if (result.hasMorePages()) {
                return result.fetchNextPage().thenCompose(this::apply).toCompletableFuture();
            }
            return CompletableFuture.completedFuture(null);
        }};

        return CompletableFuture.allOf(results.stream()
                        .map(compute::apply).toArray(CompletableFuture[]::new))
                .thenCompose(res -> {
                    int i = 0;
                    BoundStatement boundStatement =
                            getInsertOverallPS(summaryTable, rollup.rollupLevel()).bind()
                                    .setString(i++, rollup.agentRollupId())
                                    .setString(i++, query.transactionType())
                                    .setInstant(i++, Instant.ofEpochMilli(query.to()))
                                    .setDouble(i++, totalDurationNanos.doubleValue())
                                    .setDouble(i++, totalCpuNanos.doubleValue())
                                    .setDouble(i++, totalAllocatedBytes.doubleValue())
                                    .setLong(i++, transactionCount.get())
                                    .setInt(i++, rollup.adjustedTTL().generalTTL());
                    return session.writeAsync(boundStatement).toCompletableFuture();
                });
    }
```

Here are some explanations:

* the return value `ListenableFuture` is replaced with `CompletableFuture`. After this commit, the use of `ListenableFuture` from guava is almost fully removed;
* we cannot anymore have an `Iterable<Row>` as parameter since fetching rows could cause IO blocking (that's the reason why AsyncResultSet has been created). So I replaced `Iterable<Row>` with `List<AsyncResultSet>`;
* paging must be done differently now. It could not be just a for loop as previously. The new way to perform paging is described [here](https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/paging/#asynchronous-paging). I have to use an anonymous Function class instead of a lambda because of the `this::apply` call would have not worked without a class;
* since fetching the rows is performed in callbacks, the fetching will happen now in different threads. So we have to use thread-safe data structures. That's the reason why long -> AtomicLong, double -> DoubleAccumulator;
* finally, we need to wait until all the fetching is done to perform the write. It can be done using either CompletableFuture.allOf() (as in the example above) or a utility class from an external library

```
    <dependency>
      <groupId>com.spotify</groupId>
      <artifactId>completable-futures</artifactId>
      <version>0.3.1</version>
    </dependency>
```
namely, `com.spotify.futures.CompletableFutures.allAsList()`. Double-checking the differences between these two methods is on my todo. Maybe could we get rid of this external lib. I used it because it seems to have the same behavior as the `allAsList()` method used with `ListenableFuture`.

I hope my explanations are clear enough to understand why there's so much changes introduced by my PR....

So I think, the way the new cassandra driver is designed could enable glowroot to fully leverage non-blocking IO. By returning CompletionStage and / or CompletableFuture and chaining them with `thenCompose()`, we could remove all blocking get() calls and achieve better scalability. 

Just have a look at the impact it has on the CollectorServiceImpl class, grpc callbacks fit perfectly with CompletableFutures:

           ```
 aggregateDao.store(postV09AgentId, captureTime, aggregatesByTypeList, sharedQueryTexts)
                    .whenComplete((res, t) -> {
                        if (t != null) {
                            logger.error("{} - {}", postV09AgentId, t.getMessage(), t);
                            responseObserver.onError(t);
                            return;
                        }
                        // some other stuff performed here that could be chained as well....

                        responseObserver.onNext(AggregateResponseMessage.newBuilder()
                                .setNextDelayMillis(getNextDelayMillis())
                                .build());
                        responseObserver.onCompleted();
                    });
```

One more time, please note this is still a work in progress PR. It is not fully tested (far from it, though it compiles and pass all unit tests). But it gives you an overview of what I'm doing right now.

Please let me know if you feel I'm going in the wrong direction. 
That's the reason why I've added all this refactoring in a different commit so you can still consider my work on cassandra driver migration without this major refactoring of the asynchronous part.

Nowheresly commented:
I updated this PR to replace `CompletableFuture.allOf()` with `com.spotify.futures.CompletableFutures.allAsList()`.
Both are doing the same (ie wait until all `CompletableFuture` are done or fail as soon as one of them fail) except that `CompletableFuture.allOf()` return a `CompletableFuture<Void>` while the spotify version returns a `CompletableFuture<List<?>>` with a list containing the results for each `CompletableFuture`.
I think the overhead should be negligible so why not using the most exhaustive version.

Nowheresly commented:
I realized that it remains very few usage of `ListenaleFuture` after the cassandra v4 driver migration. So I did the extra step (in a separate commit, so we can choose not to merge it) to replace all these remaining usages with `CompletableFuture`.
There's nothing wrong with ListenableFuture, but I think it makes sense to try to be as consistent as possible.

Nowheresly commented:
Current status of this PR: glowroot starts and seems to work without issue/exception in non-clustered mode (I did not test yet the clustered mode), at least as far as the happy path is concerned.

I will try to focus on the cassandra driver settings and try to test the behavior for some non-happy paths.

ReillyBrogan commented:
Is this required for Glowroot to use a Cassandra 4.x backend cluster?

Nowheresly commented:
No, upgrading to the latest java driver v4 is not mandatory to use a Cassandra Cluster 4.x. But it helps to achieve a better usage (new features, perf, scalability) of the Cassandra Server v4. 

See below for details.

Java Cassandra Driver v4 is able to interact with Cassandra Server from version 2.1 to the latest 4.x.
Java Cassandra Driver v3 is able to interact with Cassandra Server from version 1.2.x to the latest 4.x as well.

To understand the difference, you need to compare the compatibility matrix for both drivers:

* driver v3 https://docs.datastax.com/en/developer/java-driver/3.7/manual/native_protocol/#compatibility-matrix
* driver v4 https://docs.datastax.com/en/developer/java-driver/4.14/manual/core/native_protocol/#compatibility-matrix

Java Driver v3 will never support native protocol v5. So, with a Java Driver v3, it possible to interact with a Cassandra Server 4.x but with a downgraded version (v4, negociated between the java driver and the cassandra server at the beginning of the session) of the native protocol.
